frontend/lib/resolution/disambiguation.cpp

```diff
@@ -1213,8 +816,215 @@ computeIsMoreVisible(Context* context,
+
+  // TODO: maybe implement the verify procedure in dyno
+  // if (fVerify) {
+  //   // check that 'compareSpecificity' creates a partial order
+  //   PartialOrderChecker checker(candidates, DC);
+
+  //   for (int i = 0; i < candidates.n; ++i) {
+  //     ResolutionCandidate* candidate1         = candidates.v[i];
+  //     bool forGenericInit = candidate1->fn->isInitializer() || candidate1->fn->isCopyInit();
+
+  //     for (int j = i; j < candidates.n; ++j) {
+  //       ResolutionCandidate* candidate2 = candidates.v[j];
+
+  //       int cmp = compareSpecificity(candidate1,
+  //                                    candidate2,
+  //                                    DC,
+  //                                    i,
+  //                                    j,
+  //                                    forGenericInit);
+
+  //       checker.addResult(i, j, cmp);
+  //     }
+  //   }
+
+  //   checker.checkResults();
+  // }
+
+}
 
 
+static MostSpecificCandidates
+disambiguateByMatch(DisambiguationContext& dctx, CandidatesVec& candidates) {
+  CandidatesVec ambiguous;
+
+  DisambiguationCandidate* bestValue = nullptr;
+
+  DisambiguationCandidate* best = (DisambiguationCandidate*)disambiguateByMatch(dctx, candidates, true,
+                                                    ambiguous);
+  MostSpecificCandidates ret = MostSpecificCandidates::getAmbiguous();
+
+  // The common case is that there is no ambiguity because the
+  // return intent overload feature is not used.
+  if (best != nullptr) {
+    ret = MostSpecificCandidates::getOnly(best->toMostSpecificCandidate(dctx.context));
+
+  } else {
+    // Now, if there was ambiguity, find candidates with different
+    // return intent in ambiguousCandidates.
+    // If there is only one of each, we are good to go.
+    int                  nRef              = 0;
+    int                  nConstRef         = 0;
+    int                  nValue            = 0;
+    int                  nOther            = 0;
+    int                  total             = 0;
+
+    DisambiguationCandidate* refCandidate      = NULL;
+    DisambiguationCandidate* constRefCandidate = NULL;
+    DisambiguationCandidate* valueCandidate    = NULL;
+
+    // Count number of candidates in each category.
+    countByReturnIntent(dctx, ambiguous, nRef, nConstRef, nValue, nOther);
+
+    total = nRef + nConstRef + nValue + nOther;
+
+    // 0 or 1 matches -> return now, not a ref pair.
+    if (total <= 1) {
+      ret = MostSpecificCandidates::getAmbiguous();
+
+      // 1 match is possible with best==NULL in cases
+      // where the 'more specific' relation is not transitive.
+
+    } else if (nOther > 0) {
+      ambiguous.clear();
+
+      // If there are *any* type/param candidates, we need to cause ambiguity
+      // if they are not selected... including consideration of where clauses.
+      bestValue  = (DisambiguationCandidate*)disambiguateByMatch(dctx, candidates, false, ambiguous);
```

Comment:
AFAIK, this is not documented in the spec https://chapel-lang.org/docs/language/spec/procedures.html#determining-most-specific-functions but it should be

Reply:
I created https://github.com/Cray/chapel-private/issues/5708 to address this need